### PR TITLE
Fix the preview crash resulted from the table view scrolling quickly.

### LINKED ISSUE #3062: Crash previewing a step from the step table — RouteLegProgress.init(leg:stepIndex:spokenInstructionIndex:)
I'm running into a crasher while running on Nav SDK betas. We should be able to update SDKs to latest in our next sprint, but wanted to share in case this helped find an issue in the SDKs

**Mapbox Navigation SDK version:**

-v2 beta 8

### Steps to reproduce

 1. Begin Navigation
 2. Tap top banner to show steps of the directions
 3. Scroll to the bottom, tap arrive at destination (sometime crash)

### Expected behavior

Not crash

### Actual behavior

Crash

stepCrash

```
Crashed: com.apple.main-thread
0  Straightaway                   0x102e7dc08 specialized RouteLegProgress.init(leg:stepIndex:spokenInstructionIndex:) + 4367932424 (RouteLegProgress.swift:4367932424)
1  Straightaway                   0x102e7c5f8 RouteLegProgress.init(leg:stepIndex:spokenInstructionIndex:) + 4367926776 (<compiler-generated>:4367926776)
2  Straightaway                   0x102d58d44 specialized NavigationViewController.topBanner(_:didSelect:stepIndex:cell:) + 960 (NavigationViewController.swift:960)
3  Straightaway                   0x102d8dbb8 protocol witness for StepsViewControllerDelegate.stepsViewController(_:didSelect:stepIndex:cell:) in conformance TopBannerViewController + 4366949304 (<compiler-generated>:4366949304)
4  Straightaway                   0x102d79af8 StepsViewController.tableView(_:didSelectRowAt:) + 206 (StepsViewController.swift:206)
5  Straightaway                   0x102d79c14 @objc StepsViewController.tableView(_:didSelectRowAt:) + 4366867476 (<compiler-generated>:4366867476)
6  UIKitCore                      0x184ea3f3c -[UITableView _selectRowAtIndexPath:animated:scrollPosition:notifyDelegate:isCellMultiSelect:] + 1104
7  UIKitCore                      0x184ea3ad4 -[UITableView _selectRowAtIndexPath:animated:scrollPosition:notifyDelegate:] + 108
8  UIKitCore                      0x184ea42a0 -[UITableView _userSelectRowAtPendingSelectionIndexPath:] + 268
9  UIKitCore                      0x185153fb8 -[_UIAfterCACommitBlock run] + 64
10 UIKitCore                      0x184cb9a68 _runAfterCACommitDeferredBlocks + 296
11 UIKitCore                      0x184ca8f98 _cleanUpAfterCAFlushAndRunDeferredBlocks + 200
12 UIKitCore                      0x184cda568 _afterCACommitHandler + 76
13 CoreFoundation                 0x1823c087c __CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__ + 32
14 CoreFoundation                 0x1823baf50 __CFRunLoopDoObservers + 604
15 CoreFoundation                 0x1823bb498 __CFRunLoopRun + 960
16 CoreFoundation                 0x1823baba0 CFRunLoopRunSpecific + 572
17 GraphicsServices               0x1990f8598 GSEventRunModal + 160
18 UIKitCore                      0x184caa3d8 -[UIApplication _run] + 1052
19 UIKitCore                      0x184caf958 UIApplicationMain + 164
20 Straightaway                   0x102a12884 main + 12 (CheckboxToggleStyle.swift:12)
21 libdyld.dylib                  0x182099568 start + 4
```

### PR DESCRIPTION
**Pull request by ShanMa1991**
### Description
This pr is to fix #3062 and #3234 by adding the check when we select the banner instruction for preview.

### Implementation
The crash reported from #3062 , #3234 and a recent crash from user are all resulted from the call of the following:
https://github.com/mapbox/mapbox-navigation-ios/blob/8257c395745e18255e391978e387034f00086290/Sources/MapboxNavigation/StepsViewController.swift#L192-L206
And then we need to form a `RouetLegProgress` to preview the step as:
https://github.com/mapbox/mapbox-navigation-ios/blob/8257c395745e18255e391978e387034f00086290/Sources/MapboxNavigation/NavigationViewController.swift#L1104-L1106

But the issue is that if we scroll the table quickly, during when the `routeProgress` changed and reload the table view data, or the `routeProgress` received in the table view is different from the `routeProgress` in the `navigationService`, there's a chance that the `stepIndex` would be out of range of the current chosen leg steps count. 

So fat the safest way is to add check for the `stepIndex` and the steps count in both the table view select and before the initialization `RouetLegProgress` for preview.

### CODE REVIEW

File: Sources/MapboxNavigation/StepsViewController.swift [hunk 1]
```diff
@@ -192,18 +192,24 @@ public class StepsViewController: UIViewController {
 extension StepsViewController: UITableViewDelegate {
     public func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
         tableView.deselectRow(at: indexPath, animated: true)
+        let legIndex = indexPath.section
         let cell = tableView.cellForRow(at: indexPath) as! StepTableViewCell
         // Since as we progress, steps are removed from the list, we need to map the row the user tapped to the actual step on the leg.
         // If the user selects a step on future leg, all steps are going to be there.
         var stepIndex: Int
-        if indexPath.section > 0 {
```
Comment (MaximAlien) [inline Sources/MapboxNavigation/StepsViewController.swift:199]: Do you think that it'd be possible to extract this logic into separate function and add unit-test for it? With unit-test we can easily provide negative test case scenario, which will help in future.

**Comment by S2Ler** [comment]
> the routeProgress received in the table view is different from the routeProgress in the navigationService, there's a chance that the stepIndex would be out of range of the current chosen leg steps count.

Why are they getting out of sync? While checking for out of bound access will fix the crash it can result in instructions to be incorrect. I'm currently working on [this](https://github.com/mapbox/mapbox-navigation-ios/tree/feature/restore-tests) branch that will fix some of the crashes related to local route progress isn't in sync with what NavNative reports. We shall investigate why top banner and navigation service are not in sync.

**Comment by ShanMa1991** [comment]
> While checking for out of bound access will fix the crash it can result in instructions to be incorrect.

No, this check is to make sure that the preview is not working, it would not result in the wrong instruction preview or the wrong top banner behavior. I know the root fix is to fix the out of sync behavior of `routeProgress` in the navigation components with the Nav `status`, but this situation may still happen when we scroll or swipe too fast or hard while the new `routeProgress` is forming. 

I think we could right now fix it for stability reason and to avoid the crashes. There're repeated crash reports from the customers caused by this issue. When we fix the out of sync issue, we could remove this safe check. Because these checks only work to prevent the instruction preview, it would not lead to wrong instruction banner presenting or any other weird behavior.

File: Sources/MapboxNavigation/StepsViewController.swift (hunk 1 above)
Comment (ShanMa1991) [reply Sources/MapboxNavigation/StepsViewController.swift:199]: Thanks for the suggestion! I'll add test for this part.

File: Sources/MapboxNavigation/NavigationViewController.swift [hunk 2]
```diff
@@ -1103,6 +1103,7 @@ extension NavigationViewController: TopBannerViewControllerDelegate {
     
     public func topBanner(_ banner: TopBannerViewController, didSelect legIndex: Int, stepIndex: Int, cell: StepTableViewCell) {
         let progress = navigationService.routeProgress
+        guard stepIndex < progress.route.legs[legIndex].steps.count else { return }
```
Comment (bamx23) [inline Sources/MapboxNavigation/NavigationViewController.swift:1106]: Should we also check `legIndex` here? Or to use `safe:` subscript?

Comment (ShanMa1991) [reply Sources/MapboxNavigation/NavigationViewController.swift]: Thanks for the suggestion, I'll add check to this part as well.

File: Sources/MapboxNavigation/StepsViewController.swift [hunk 3]
```diff
@@ -192,18 +192,21 @@ public class StepsViewController: UIViewController {
...
+        if legIndex > 0 {
             stepIndex = indexPath.row
         } else {
             stepIndex = indexPath.row + routeProgress.currentLegProgress.stepIndex
             // For the current leg, we need to know the upcoming step.
-            stepIndex += indexPath.row + 1 > sections[indexPath.section].count ? 0 : 1
+            stepIndex += indexPath.row + 1 > sections[legIndex].count ? 0 : 1
```
Comment (1ec5) [inline Sources/MapboxNavigation/StepsViewController.swift:205]:
```suggestion
            if sections[legIndex].indices.contains(indexPath.row + 1) {
                stepIndex += 1
            }
```

**Review by 1ec5 — APPROVED** [review]

File: Sources/MapboxNavigation/Route.swift [hunk 4]
```diff
@@ -155,6 +155,16 @@ extension Route {
...
     }
+    
+    /**
+     Returns true if both the legIndex and stepIndex are valid in the route.
+     */
+    func checkSelectedIndex(legIndex: Int, stepIndex: Int) -> Bool {
+        guard let stepCount = legs[safe: legIndex]?.steps.count,
+              stepIndex < stepCount,
+              stepIndex >= 0 else { return false }
+        return true
```
[the remark this answers is not in the document]
Comment (ShanMa1991) [reply Sources/MapboxNavigation/Route.swift:166]: Thanks for the suggestion, I'll improve this part.
